replace_with/src/lib.rs

```diff
@@ -0,0 +1,8 @@
+use std::ptr;
+
+pub fn replace_with<T, F: FnOnce(T) -> T>(t: &mut T, f: F) {
+    let p = t as *mut T;
+    unsafe {
+        ptr::write(p, f(ptr::read(p)))
```

Comment:
Nice!

Ross discovered that if you use `t` in place of `p` in the function calls, then the `&mut` will be coerced to the appropriate pointer type. This tells us that function calls trigger coercion on the arguments.